File: Algorithms/Math/60. Permutation Sequence.py
class Solution(object):
    def getPermutation(self, n, k):
		factorials = [1]
		for num in range(1, n+1):
			factorials.append(factorials[-1] * num)
		nums = range(1, n+1)
		res = list()
		k -= 1
		for inx in range(1, n+1):
			bit, k = divmod(k, factorials[n - inx])
			res.append(nums[bit])
			nums.pop(bit)
		return ''.join(map(str, res))

File: Algorithms/Math/60. Permutation Sequence.cpp
class Solution {
public:
    string getPermutation(int n, int k) {
        vector<int> factorial(n + 1);
        vector<int> numbers(n);
        factorial[0] = 1;
        for(int i = 1; i <= n; ++i)
        {
            numbers[i - 1] = i;
            factorial[i] = factorial[i - 1] * i; 
        }
        
        string res;
        k -= 1;
        for(int i = 1; i <= n; ++i)
        {
            int index = k / factorial[n - i];
            res.push_back('0' + numbers[index]);
            numbers.erase(numbers.begin() + index);
            k %= factorial[n - i];
        }        
        return res;     
    }
};
